loki/transformations/single_column/vertical.py

```diff
@@ -0,0 +1,218 @@
+        to demote!
+        """
+        loop_var_map = {}
+        with pragmas_attached(routine, ir.Loop):
+            for loop in FindNodes(ir.Loop).visit(routine.body):
+                if loop.variable.name.lower() == self.vertical.index.lower():
+                    ignore = False
+                    if is_loki_pragma(loop.pragma, starts_with='fused-loop'):
+                        parameters = get_pragma_parameters(loop.pragma, starts_with='fused-loop')
+                        group = parameters.get('group', 'default')
+                        if group == 'ignore':
+                            ignore = True
+                    if not ignore:
+                        loop_var_map[loop] = as_tuple(var.name.lower() for var in FindVariables().visit(loop.body)
+                                if isinstance(var, sym.Array))
+
+        safe_to_demote = ()
+        for var in demote_candidates:
+            count = 0
+            for _, var_names in loop_var_map.items():
+                if var in var_names:
+                    count += 1
+            if count <= 1:
+                safe_to_demote += (var,)
+
+        return safe_to_demote
+
+    def find_relevant_local_arrays(self, routine):
+        """
+        Find local arrays/temporaries that do have the vertical dimension.
+        """
+        # local/temporary arrays
+        arrays = [var for var in FindVariables(unique=False).visit(routine.body) if isinstance(var, sym.Array)]
+        argument_names = [arg.name.lower() for arg in routine.arguments]
+        local_arrays = [arr for arr in arrays if arr.name.lower() not in argument_names]
+        # only those with the vertical size within shape
+        relevant_local_arrays = [arr for arr in local_arrays if self.vertical.size.lower()
+                in [var.name.lower() for var in FindVariables().visit(arr.shape)]]
+        # filter arrays to be ignored (for whatever reason)
+        ignore_names = self.find_local_arrays_to_be_ignored(routine)
+        if ignore_names:
+            relevant_local_arrays = [arr for arr in relevant_local_arrays if arr.name.lower() not in ignore_names]
+        return relevant_local_arrays
+
+    def find_local_arrays_to_be_ignored(self, routine):
+        """
+        Identify variables to be ignore regarding demotion for whatever reason.
+
+        Reasons are:
+
+        * explicitly marked to be ignored via pragmas within the input source file, e.g.,
+          'loki k-caching ignore(var1, var2, ...)'
+        """
+        ignore = ()
+        pragmas = FindNodes(ir.Pragma).visit(routine.body)
+        # look for 'loki k-caching ignore(var1, var2, ...)' pragmas within routine and ignore those vars
+        for pragma in pragmas:
+            if is_loki_pragma(pragma, starts_with='k-caching'):
+                pragma_ignore = get_pragma_parameters(pragma, starts_with='k-caching').get('ignore', None)
```

Comment:
Minor nitpick: I wonder if the `is_loki_pragma` and `get_pragma_parameters` can be combined to something like `if pragma_ignore := get_pragma_parameters(pragma, starts_with='k-caching').get(...)`. Not strictly necessary, just wondering...